### PR TITLE
create a Orocos::Async.proxy Object for websocket reads in order to r…

### PR DESCRIPTION
**Pull request by planthaber**
…eally connect port on being available

Looks like async ports are only working with an proxy object.

### CODE REVIEW

File: lib/rock/webapp/tasks/api.rb [hunk 1]
```diff
@@ -155,24 +155,22 @@ def get_operation(name_service, name, operation_name)
...
-    
-                        if !port.is_reader?
-                            error! "#{port.name} is an input port, cannot read"
-                        end
-                        
                         if Faye::WebSocket.websocket?(env)
-                            port = port.port.to_async.reader(init: true, pull: true)
+                            proxy = Orocos::Async.proxy("#{params[:name_service]}/#{params[:name]}")
+                            asyncport = proxy.port(params[:port_name]);
                             count = params.fetch(:count, Float::INFINITY)
```
Comment (planthaber) [inline lib/rock/webapp/tasks/api.rb:180]: Is there a way to check the async port for a "reader", respond_to?(:reader) does not work.

**Comment by doudou** [comment]
This changes the semantic of the method. The current semantic is to return an error if the port does not exist, which is something I really want to be able to rely on. If you want a proxy-like behaviour, then create another endpoint with that semantic, or add a parameter to that effect.

**Comment by planthaber** [comment]
True, made it optional

**Comment by planthaber** [comment]
rebased and squashed (proxy is optional and not default, added hostname regex)

File: lib/rock/webapp/tasks/api.rb [hunk 2]
```diff
@@ -156,24 +156,32 @@ def get_operation(name_service, name, operation_name)
...
-                            error! "#{port.name} is an input port, cannot read"
-                        end
 
+                    get ':name_service/:name/ports/:port_name/read', requirements: { name_service: ValidHostnameRegex } do
+                    params do
+                        oprional :proxy, type: Boolean, default: false
+                    end
                         if Faye::WebSocket.websocket?(env)
-                            port = port.port.to_async.reader(init: true, pull: true)
+                            port = nil
```
Comment (doudou) [inline lib/rock/webapp/tasks/api.rb]: You don't need to assign the variable here.

**Comment by planthaber** [comment]
removed assignment, rebased and sqashed

File: lib/rock/webapp/tasks/api.rb [hunk 3]
```diff
@@ -156,24 +156,31 @@ def get_operation(name_service, name, operation_name)
...
-                        port = get_port(*params.values_at('name_service', 'name', 'port_name', 'init', 'timeout'))
-                        #port = port_by_task_and_name(*params.values_at('name_service', 'name', 'port_name'))
-
-                        if !port.is_reader?
-                            error! "#{port.name} is an input port, cannot read"
-                        end
 
+                    get ':name_service/:name/ports/:port_name/read', requirements: { name_service: ValidHostnameRegex } do
+                    params do
+                        oprional :proxy, type: Boolean, default: false
```
Comment (doudou) [inline lib/rock/webapp/tasks/api.rb]: You have not even loaded the library, or ?

**Comment by planthaber** [comment]
Now it works

File: lib/rock/webapp/tasks/api.rb [hunk 4]
```diff
@@ -156,24 +156,31 @@ def get_operation(name_service, name, operation_name)
...
-                        end
 
+                    get ':name_service/:name/ports/:port_name/read', requirements: { name_service: ValidHostnameRegex } do
+                    params do
+                        optional :proxy, type: Boolean, default: false
+                    end
                         if Faye::WebSocket.websocket?(env)
-                            port = port.port.to_async.reader(init: true, pull: true)
+                            if params["proxy"]
+                                proxy = Orocos::Async.proxy("#{params[:name_service]}/#{params[:name]}")
```
Comment (doudou) [inline lib/rock/webapp/tasks/api.rb]: Wouldn't this fail with '*' as name service ?

Comment (planthaber) [reply lib/rock/webapp/tasks/api.rb]: Tested it manually, it actually works.

I guess the Async.proxy constructor supports the *

Comment (doudou) [reply lib/rock/webapp/tasks/api.rb]: > I guess the Async.proxy constructor supports the *

No it does not. Or at least, not on purpose (which given the sorry state of async might well be the case).

Handle the '*' explicitely here, relying on orocos.rb to do it is a recipe for disaster.

File: lib/rock/webapp/tasks/api.rb [hunk 5]
```diff
@@ -162,17 +162,24 @@ def get_operation(name_service, name, operation_name)
                         optional :proxy, type: Boolean, default: false
                     end
                         if Faye::WebSocket.websocket?(env)
-                            if params["proxy"]
-                                proxy = Orocos::Async.proxy("#{params[:name_service]}/#{params[:name]}")
-                                port = proxy.port(params[:port_name]);
+                            puts params[:proxy]
```
Comment (doudou) [inline lib/rock/webapp/tasks/api.rb]: Leftover debug code.

Comment (planthaber) [reply lib/rock/webapp/tasks/api.rb]: as always :-/

**Comment by doudou** [comment]
:+1:

**Pull request merged**